nemo/collections/nlp/modules/common/megatron/transformer.py

```diff
@@ -732,64 +845,77 @@ def forward(
         # dropout semantics during training and inference phases.
-        if self.bias and self.bias_dropout_fusion:
-            if self.training:
-                bias_dropout_add_func = bias_dropout_add_fused_train
-            else:
-                bias_dropout_add_func = bias_dropout_add_fused_inference
-        elif self.bias and not self.bias_dropout_fusion:
-            bias_dropout_add_func = get_bias_dropout_add(self.training)
-        else:
-            bias_dropout_add_func = dropout_add
+
+        bias_dropout_add_func = self._get_bias_droput_add_func(arch=self.arch, position_after='attention')
 
         if attention_bias is not None:
             attention_bias = attention_bias.expand_as(residual)
 
         layernorm_input = bias_dropout_add_func(attention_output, attention_bias, residual, self.hidden_dropout)
 
         # Layer norm post the self attention.
-        layernorm_output = self.post_attention_layernorm(layernorm_input)
+        normalization_output = self.post_attention_normalization(layernorm_input)
 
         if self.layer_type == LayerType.decoder:
             attention_output, attention_bias = self.inter_attention(
-                layernorm_output, enc_dec_attn_mask, encoder_output=encoder_output
+                normalization_output, enc_dec_attn_mask, encoder_output=encoder_output
             )
+            # If normformer, apply norm on the output of the self attention.
+            if self.arch == 'normformer':
+                # Normformer normalization
+                attention_output = (
+                    attention_output + attention_bias if attention_bias is not None else attention_output
+                )
+                attention_output = self.post_inter_attention_normformer_norm(attention_output)
+                attention_bias = None
+
             # residual connection
             if self.apply_residual_connection_post_layernorm:
-                residual = layernorm_output
+                residual = normalization_output
             else:
                 residual = layernorm_input
 
             if attention_bias is not None:
                 attention_bias = attention_bias.expand_as(residual)
 
+            bias_dropout_add_func = self._get_bias_droput_add_func(arch=self.arch, position_after='attention')
+
             layernorm_input = bias_dropout_add_func(attention_output, attention_bias, residual, self.hidden_dropout)
 
             # Layer norm post the decoder attention
-            layernorm_output = self.post_inter_attention_layernorm(layernorm_input)
+            normalization_output = self.post_inter_attention_normalization(layernorm_input)
 
         # MLP.
-        mlp_output, mlp_bias = self.mlp(layernorm_output)
+        mlp_output, mlp_bias = self.mlp(normalization_output)
 
         # Second residual connection.
         if self.apply_residual_connection_post_layernorm:
-            residual = layernorm_output
+            residual = normalization_output
```

Comment:
`residual` for FFN should be scaled by a learnable parameter `lambda_resid`?
![image](https://user-images.githubusercontent.com/43824965/165166908-67f833f7-1c9c-4c88-9735-fb062284ceb8.png)

Reply:
The paper shows that it hurts perplexity so I didn't add it (see Fig.2 in their paper)